tpke/src/combine.rs

```diff
@@ -2,71 +2,100 @@
             E::G2Prepared::from(
-                izip!(
-                    lagrange_slice.iter(),
-                    decrypter.blinded_key_shares.blinded_key_shares.iter() //decrypter.blinded_key_shares.window_tables.iter()
-                )
-                .map(|(lambda, blinded_key_share)| {
-                    blinded_key_share.mul(*lambda)
-                })
-                /*.map(|(lambda, base_table)| {
-                    FixedBaseMSM::multi_scalar_mul::<E::G2Projective>(
-                        scalar_bits,
-                        window_size,
-                        &base_table.window_table,
-                        &[*lambda],
-                    )[0]
-                })*/
-                .sum::<E::G2Projective>()
-                .into_affine(),
+                // [b]Z_i * L_i
+                blinded_key_share.mul(*lambda).into_affine(),
             )
         })
         .collect::<Vec<_>>()
 }
+pub fn prepare_combine_simple<E: PairingEngine>(
+    shares_x: &[E::Fr],
+) -> Vec<E::Fr> {
+    // Calculate lagrange coefficients using optimized formula, see https://en.wikipedia.org/wiki/Lagrange_polynomial#Optimal_algorithm
+    let mut lagrange_coeffs = vec![];
+    for x_j in shares_x {
+        let mut prod = E::Fr::one();
+        for x_m in shares_x {
+            if x_j != x_m {
+                // In this formula x_i = 0, hence numerator is x_m
+                prod *= (*x_m) / (*x_m - *x_j);
+            }
+        }
+        lagrange_coeffs.push(prod);
+    }
+    lagrange_coeffs
+}
 
 pub fn share_combine<E: PairingEngine>(
     shares: &[DecryptionShare<E>],
     prepared_key_shares: &[E::G2Prepared],
 ) -> E::Fqk {
     let mut pairing_product: Vec<(E::G1Prepared, E::G2Prepared)> = vec![];
 
-    for (d_i, blinded_key_share) in izip!(shares, prepared_key_shares.iter()) {
-        // e(D_i, [b*omega_i^-1] Z_{i,omega_i})
+    for (d_i, prepared_key_share) in izip!(shares, prepared_key_shares.iter()) {
+        // e(D_i, [b*omega_i^-1] Z_{i,omega_i}), TODO: Is this formula correct?
         pairing_product.push((
+            // D_i
             E::G1Prepared::from(d_i.decryption_share),
-            blinded_key_share.clone(),
+            // Z_{i,omega_i}) = [dk_{i}^{-1}]*\hat{Y}_{i_omega_j}]
+            // Reference: https://nikkolasg.github.io/ferveo/pvss.html#validator-decryption-of-private-key-shares
+            // Prepared key share is a sum of L_i * [b]Z_i
+            prepared_key_share.clone(),
         ));
     }
     E::product_of_pairings(&pairing_product)
 }
 
+pub fn share_combine_simple<E: PairingEngine>(
+    shares: &[E::Fqk],
+    lagrange_coeffs: &[E::Fr],
+) -> E::Fqk {
+    let mut product_of_shares = E::Fqk::one();
+
+    // Sum of C_i^{L_i}z
+    for (c_i, alpha_i) in zip_eq(shares.iter(), lagrange_coeffs.iter()) {
+        // Exponentiation by alpha_i
+        let ss = c_i.pow(alpha_i.into_repr());
+        product_of_shares *= ss;
+    }
```

Comment:
Not for this PR, but we need to explore a potential optimization where the exponentiation is performed before the pairing (which implies that either the request does it before the request, or `alpha_i` is shared with the node and the node does it), and in the combination side we use something for products of pairings.